### PR TITLE
mount.ceph: remove `ms_mode' mount option when switching to old-syntax

### PR DESCRIPTION
**Pull request by vshankar**
... if the user hasn't specified using one explicity. kernel versions
< 5.11 do not understand `ms_mode' mount option which would result in
mount failures.

Fixes: http://tracker.ceph.com/issues/55110
Signed-off-by: Venky Shankar <vshankar@redhat.com>

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

**Comment by jtlayton** [comment]
This fix looks fine for now, but I do wonder whether it might be better to not add on ms_mode in the mount helper, and have the kernel instead default to using ms_mode=prefer-crc when the new syntax is used and no other ms_mode value is specified. Trying to manage sane defaults for multiple kernel versions in the mount helper is harder to get right (as we're finding out).

**Review by ajarr — APPROVED** [review]

**Review by lxbsz — APPROVED** [review]

**Review by batrick — APPROVED** [review]

**Comment by vshankar** [comment]
jenkins test make check

**Comment by vshankar** [comment]
@jtlayton please take a look.

**Review by jtlayton — APPROVED** [review]

**Comment by vshankar** [comment]
@jtlayton please take a look. I had to change the logic a bit to suit all combinations. Notable changes:

- start by fetching v2 addresses by default (unless ms_mode = legacy is used)
- fallback to v1 addresses if no v2 addresses are available and ms_mode is _not_ explicitly specified
- when falling back to old-syntax, switch to v1 addresses if ms_mode is _not_ explicitly specified

With the above changes, there are no kclient related mount failures with fs suite tests, however, I am still unable to get a clean fs suite run due to teuthology and other infra noise.

I'll work on resolving them and getting a clean run (at-least for kclient related jobs).

**Review by jtlayton — APPROVED** [review]

**Comment by vshankar** [comment]
There was a bug where the mount helper would reinitialize cmi->mon_addr with v1 addresses even if mon_addrs were provided via v1-syntax. This was causing `xfstests generic/099' to fail since it expect the device in /proc/mounts to match its provided dev used in mount syntax.

@lxbsz please take a look.

FYI, here are the runs: https://pulpito.ceph.com/?branch=wip-vshankar-testing-55110-20220408-203242

- fs suite failed are unrelated, re-running failed/dead jobs
- just scheduled smoke tests

File: src/mount/mount.ceph.c [hunk 1]
```diff
@@ -323,8 +325,12 @@ static int parse_dev(const char *dev_str, struct ceph_mount_info *cmi,
 	return ret;
 }
 
-/* resolve monitor host and record in option string */
-static int finalize_src(struct ceph_mount_info *cmi, int *opt_pos)
+/* resolve monitor host and optionallyrecord in option string.
```
Comment (lxbsz) [inline src/mount/mount.ceph.c:328]: s/optionallyrecord/optionally record/

Comment (vshankar) [reply src/mount/mount.ceph.c]: done

**Comment by vshankar** [comment]
jenkins test make check

**Comment by vshankar** [comment]
https://tracker.ceph.com/projects/cephfs/wiki/Master#2022-Apr-11

**Pull request merged**